File: app/src/lib/server/indexer.ts
import { db, schema } from "./db/index.js";
import { eq } from "drizzle-orm";
import { readNote, listNotes } from "./fs/vault.js";
import { parseMarkdownAst } from "@prometheus/core/markdown";
import { extractWikilinks } from "@prometheus/core/markdown";
import { extractTasks } from "@prometheus/core/markdown";

export async function indexNote(notePath: string): Promise<void> {
  const note = await readNote(notePath);

  // Check if note already indexed with same checksum
  const existing = await db
    .select({ checksum: schema.notes.checksum })
    .from(schema.notes)
    .where(eq(schema.notes.id, note.frontmatter.id))
    .limit(1);

  if (existing.length > 0 && existing[0].checksum === note.checksum) {
    return; // No changes
  }

  const ast = parseMarkdownAst(note.content);
  const wikilinks = extractWikilinks(ast);
  const tasks = extractTasks(ast);

  // Delete existing entry for this path (handles ID changes)
  await db.delete(schema.notes).where(eq(schema.notes.path, note.path));

  // Insert note
  await db
    .insert(schema.notes)
    .values({
      id: note.frontmatter.id,
      path: note.path,
      title: note.frontmatter.title,
      content: note.content,
      createdAt: new Date(note.frontmatter.created),
      modifiedAt: new Date(note.frontmatter.modified),
      theme: note.frontmatter.theme ?? null,
      checksum: note.checksum,
    })
    .onConflictDoUpdate({
      target: schema.notes.id,
      set: {
        path: note.path,
        title: note.frontmatter.title,
        content: note.content,
        modifiedAt: new Date(note.frontmatter.modified),
        theme: note.frontmatter.theme ?? null,
        checksum: note.checksum,
      },
    });

  // Re-index links
  await db
    .delete(schema.links)
    .where(eq(schema.links.sourceId, note.frontmatter.id));

  if (wikilinks.length > 0) {
    await db.insert(schema.links).values(
      wikilinks.map((wl) => ({
        sourceId: note.frontmatter.id,
        targetId: null,
        targetSlug: wl.slug,
        context: wl.displayText,
      })),
    );
  }

  // Re-index tasks
  await db
    .delete(schema.tasks)
    .where(eq(schema.tasks.noteId, note.frontmatter.id));

  if (tasks.length > 0) {
    await db.insert(schema.tasks).values(
      tasks.map((t) => ({
        noteId: note.frontmatter.id,
        content: t.content,
        completed: t.completed,
        lineNumber: t.lineNumber,
      })),
    );
  }

  // Re-index tags
  await db
    .delete(schema.noteTags)
    .where(eq(schema.noteTags.noteId, note.frontmatter.id));

  if (note.frontmatter.tags.length > 0) {
    for (const tagName of note.frontmatter.tags) {
      // Upsert tag
      await db
        .insert(schema.tags)
        .values({ name: tagName })
        .onConflictDoNothing();

      const [tag] = await db
        .select({ id: schema.tags.id })
        .from(schema.tags)
        .where(eq(schema.tags.name, tagName))
        .limit(1);

      if (tag) {
        await db
          .insert(schema.noteTags)
          .values({ noteId: note.frontmatter.id, tagId: tag.id })
          .onConflictDoNothing();
      }
    }
  }
}

export async function indexAllNotes(): Promise<number> {
  const notes = await listNotes();
  const vaultPaths = new Set(notes);
  let count = 0;

  // Index all existing files
  for (const notePath of notes) {
    try {
      await indexNote(notePath);
      count++;
    } catch (e) {
      console.error(`Failed to index ${notePath}:`, e);
    }
  }

  // Remove stale entries (files that no longer exist in vault)
  const dbNotes = await db
    .select({ id: schema.notes.id, path: schema.notes.path })
    .from(schema.notes);

  for (const dbNote of dbNotes) {
    if (!vaultPaths.has(dbNote.path)) {
      await db.delete(schema.notes).where(eq(schema.notes.id, dbNote.id));
    }
  }

  return count;
}

export async function removeNoteFromIndex(notePath: string): Promise<void> {
  const [note] = await db
    .select({ id: schema.notes.id })
    .from(schema.notes)
    .where(eq(schema.notes.path, notePath))
    .limit(1);

  if (note) {
    await db.delete(schema.notes).where(eq(schema.notes.id, note.id));
  }
}

File: internal/service/indexer.go
package service

import (
	"fmt"
	"log"

	"github.com/wisteriahuman/prometheus/internal/db"
)

type Indexer struct {
	db    *db.DB
	vault *Vault
	md    *Markdown
}

func NewIndexer(database *db.DB, vault *Vault, md *Markdown) *Indexer {
	return &Indexer{db: database, vault: vault, md: md}
}

func (idx *Indexer) IndexNote(notePath string) error {
	note, err := idx.vault.ReadNote(notePath)
	if err != nil {
		return fmt.Errorf("read note %s: %w", notePath, err)
	}

	// Check if already indexed with same checksum
	var existingChecksum string
	row := idx.db.QueryRow("SELECT checksum FROM notes WHERE id = ?", note.Frontmatter.ID)
	row.Scan(&existingChecksum)
	if existingChecksum == note.Checksum {
		return nil
	}

	wikilinks := ExtractWikilinks(note.Content)
	tasks := ExtractTasks(note.Content)

	tx, err := idx.db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	// Delete existing entry for this path (handles ID changes)
	tx.Exec("DELETE FROM notes WHERE path = ?", note.Path)

	// Upsert note
	_, err = tx.Exec(`
		INSERT INTO notes (id, path, title, content, created_at, modified_at, theme, checksum)
		VALUES (?, ?, ?, ?, strftime('%s', ?), strftime('%s', ?), ?, ?)
		ON CONFLICT(id) DO UPDATE SET
			path = excluded.path,
			title = excluded.title,
			content = excluded.content,
			modified_at = excluded.modified_at,
			theme = excluded.theme,
			checksum = excluded.checksum
	`, note.Frontmatter.ID, note.Path, note.Frontmatter.Title, note.Content,
		note.Frontmatter.Created, note.Frontmatter.Modified,
		themePtr(note.Frontmatter.Theme), note.Checksum)
	if err != nil {
		return fmt.Errorf("upsert note: %w", err)
	}

	// Update FTS5 index
	tx.Exec("DELETE FROM notes_fts WHERE rowid IN (SELECT rowid FROM notes_fts WHERE title = ? OR content = ?)",
		note.Frontmatter.Title, note.Content)
	tx.Exec("INSERT INTO notes_fts (rowid, title, content) VALUES ((SELECT rowid FROM notes WHERE id = ?), ?, ?)",
		note.Frontmatter.ID, note.Frontmatter.Title, note.Content)

	// Re-index links
	tx.Exec("DELETE FROM links WHERE source_id = ?", note.Frontmatter.ID)
	for _, wl := range wikilinks {
		ctx := wl.DisplayText
		if ctx == "" {
			ctx = wl.Slug
		}
		tx.Exec("INSERT INTO links (source_id, target_id, target_slug, context) VALUES (?, NULL, ?, ?)",
			note.Frontmatter.ID, wl.Slug, ctx)
	}

	// Re-index tasks
	tx.Exec("DELETE FROM tasks WHERE note_id = ?", note.Frontmatter.ID)
	for _, t := range tasks {
		completed := 0
		if t.Completed {
			completed = 1
		}
		tx.Exec("INSERT INTO tasks (note_id, content, completed, line_number) VALUES (?, ?, ?, ?)",
			note.Frontmatter.ID, t.Content, completed, t.Line)
	}

	// Re-index tags
	tx.Exec("DELETE FROM note_tags WHERE note_id = ?", note.Frontmatter.ID)
	for _, tagName := range note.Frontmatter.Tags {
		tx.Exec("INSERT OR IGNORE INTO tags (name) VALUES (?)", tagName)

		var tagID int64
		row := tx.QueryRow("SELECT id FROM tags WHERE name = ?", tagName)
		if err := row.Scan(&tagID); err != nil {
			continue
		}
		tx.Exec("INSERT OR IGNORE INTO note_tags (note_id, tag_id) VALUES (?, ?)",
			note.Frontmatter.ID, tagID)
	}

	return tx.Commit()
}

func (idx *Indexer) IndexAll() int {
	notes, err := idx.vault.ListNotes()
	if err != nil {
		log.Printf("Failed to list notes: %v", err)
		return 0
	}

	vaultPaths := make(map[string]bool)
	for _, n := range notes {
		vaultPaths[n] = true
	}

	count := 0
	for _, notePath := range notes {
		if err := idx.IndexNote(notePath); err != nil {
			log.Printf("Failed to index %s: %v", notePath, err)
			continue
		}
		count++
	}

	// Remove stale entries
	rows, err := idx.db.Query("SELECT id, path FROM notes")
	if err == nil {
		defer rows.Close()
		for rows.Next() {
			var id, path string
			rows.Scan(&id, &path)
			if !vaultPaths[path] {
				idx.db.Exec("DELETE FROM notes WHERE id = ?", id)
			}
		}
	}

	return count
}

func (idx *Indexer) RemoveNote(notePath string) {
	idx.db.Exec("DELETE FROM notes WHERE path = ?", notePath)
}

func themePtr(t *string) interface{} {
	if t == nil {
		return nil
	}
	return *t
}
